### PR TITLE
PR: Enhance translation methods

### LINKED ISSUE #2513: Deferred translations for Warnings and Exceptions
## 🚀 Feature

As raised by @Czaki on https://napari.zulipchat.com/#narrow/stream/212875-general/topic/Translation.20of.20exceptions

## Motivation

When using napari as an application, it makes sense to translate as much of the strings for the user, whenever they are using a different UI language.

However when using napari on scripts, we might want to preserve the error messages in the original english, for when things go wrong. To do this we can have deferred translations, so that the notifications manager will only try to translate a string when displaying it on the UI.

This poses some issues if the original string included variables, or format like:

```python
variable = "boo"
raise ValueError(f"Some message with a {variable}")
```

## Alternatives

We can use a parameter for the current `trans._` methods like:

```python
trans._("string", deferred=True)
``` 

so that we return the same string, but we can now extract it for the catalog.

```python
f"string {variable}"
```

for these cases with variables we might want to do something like:

```python
trans._("string {variable}", args={"variable": variable}, deferred=True)

or

trans._("string {variable}", deferred=True, variable=variable)
```

or similar and we will need to define a subclass of `str` that knows how to do some things for us when we want to use on the notifications manager.

## Additional context

See https://docs.python.org/3/library/gettext.html#deferred-translations

### PR DESCRIPTION
**Pull request by goanpeca**
# Description

Fixes #2513

Include format `kwargs` and a `deferred` option, for exceptions raised in modules not displaying a UI.

A subsequent PR will update the notifications UI to use deferred translations and translate before showing them on the UI.

Part of the docs thgat explain what this does.

Screen Shot 2021-04-07 at 11 44 20 AM

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Review by sofroniewn — APPROVED** [review]
Thanks for the docs update hear, and thanks I believe to @Czaki who was part of the original discussion leading to this idea.

This looks great to me, but maybe @jni @Czaki would like to comment before merge!!

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py]:
```suggestion
        msgctxt: Optional[str] = None,
```

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py]:
```suggestion
        n: Optional[int] = None,
```

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py]:
```suggestion
        domain: Optional[str] = None,
        msgctxt: Optional[str] = None,
        msgid: Optional[str] = None,
        msgid_plural: Optional[str] = None,
        n: Optional[str] = None,
```

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py]:
```suggestion
from typing import Optional, Union
```

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py:230]: I have little worries about the number of ifs in these functions.  

These two lines could be merged in one `elif` but it may break semantic code split but I would like to see something simpler here.

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py]: Use `return` instead of obsolete `translation` variable. Simplify `if` clause

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py]:
```suggestion
        deferred: Optional[bool] = False,
```

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py]: do not use `translation` variable. Use `return` statement and remove obsolete `else`.

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py]:
```suggestion
        self, msgctxt: str, msgid: str, deferred: Optional[bool] = False, **kwargs
```

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py]: instead of using `translation` variable use `return` and remove `else`.

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py]:
```suggestion
        deferred: Optional[bool] = False,
```

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py:526]: No assignment to the variable. It looks like lack of test for this PR.

File: napari/utils/translations.py
Comment (Czaki) [inline napari/utils/translations.py]: Instead use `translation` variable use `return` statement and remove `else`.

**Review by Czaki — CHANGES_REQUESTED** [review]
I like this PR. I point a few places. 

As this is the critical part of interface presentation which may produce hard to debug errors (missed text) I suggest adding full coverage with tests and check all types annotation.

File: napari/utils/translations.py
Comment (goanpeca) [reply napari/utils/translations.py:230]: Yes, probably, but keeping it separate makes it clear to understand what is going on.

File: napari/utils/translations.py
Comment (goanpeca) [reply napari/utils/translations.py:526]: Fixed and added tests!

**Comment by goanpeca** [comment]
Thanks for the review @sofroniewn and @Czaki .

---

I made the requested fixes @Czaki, except the one on the `if/if/else` comments. Added tests for all the methods, indeed I was missing some paths!

File: napari/utils/translations.py
Comment (sofroniewn) [reply napari/utils/translations.py:230]: i'm ok leaving this as its a little more readable, but thanks for flagging for extra discussion

Comment (Czaki) [reply napari/utils/translations.py:230]: maybe split this function and simplify all code using fact that there is no need for `else` after the `return` statement.

Comment (goanpeca) [reply napari/utils/translations.py:230]: I personally do not like having multiple return statements in a single function, specially if it is a longer one like this one.

So I prefer a single exit point even if it implies creating an extra variable.

```python
def something():
    if some_condition:
        var = ...
    else:
        var = ...

    return var
```

Than something like this.

```python
def something():
    if some_condition:
        return ...

    return ...
```

In my opinion, the clarity lost is not worth the couple of lines saved and having a variable that tells you what is storing instead of a complex expression on a `return` statement makes it more readable (to me). Even if there is an extra branch to process in python because of the else.

I am open to suggestions on whether there is a preferred approach in the napari codebase.

Comment (Czaki) [reply napari/utils/translations.py:230]: I personally prefer multiple `return` statements. Reading function I see that this is the end of branch and I do not need this part of code to understand the next part.

Comment (tlambert03) [reply napari/utils/translations.py:230]: This feels like a bit of a bikeshed discussion.  I find both reasonable in various cases, and lacking a general style guide for napari, I think it’s fine to leave it up to the contributor of the PR

Comment (goanpeca) [reply napari/utils/translations.py:230]: Agreed :)

**Comment by goanpeca** [comment]
Besides some coding style and preferences discussion, that could be addressed separately, I think this is ready @sofroniewn .

Thanks for the reviews @tlambert03 and @Czaki :)

Also the PR that check on CI for untranslated strings will benefit a lot from this PR getting merged (WIP https://github.com/napari/napari/pull/2521) :-p

**Comment by sofroniewn** [comment]
Ok, I think this is now ready to go in. Thanks for the reviews all and edits!!

**Pull request merged**
